src/banking_stage.rs

```diff
@@ -241,63 +239,163 @@ impl Service for BankingStage {
+
+        // send 'em over
+        let recycler = PacketRecycler::default();
+        let packets = to_packets(&recycler, &[tx, tx_no_ver, tx_anf]);
+
+        // glad they all fit
+        assert_eq!(packets.len(), 1);
+        verified_sender                       // tx, no_ver, anf
+            .send(vec![(packets[0].clone(), vec![1u8, 0u8, 1u8])])
+            .unwrap();
+
+        drop(verified_sender);
+
+        let entries: Vec<_> = entry_receiver.iter().map(|x| x).collect();
+        assert_eq!(entries.len(), 1);
+        let mut last_id = start_hash;
+        entries.iter().for_each(|entries| {
+            assert_eq!(entries.len(), 1);
+            assert!(entries.verify(&last_id));
+            last_id = entries.last().unwrap().id;
+        });
+        assert_eq!(banking_stage.join().unwrap(), ());
+    }
+
+    #[test]
+    fn test_banking_stage_entryfication() {
+        // In this attack we'll demonstrate that a verifier can interpret the ledger
+        // differently if either the server doesn't signal the ledger to add an
+        // Entry OR if the verifier tries to parallelize across multiple Entries.
+        let mint = Mint::new(2);
+        let bank = Bank::new(&mint);
+        let recycler = PacketRecycler::default();
+        let (verified_sender, verified_receiver) = channel();
+        let (banking_stage, entry_receiver) =
+            BankingStage::new(Arc::new(bank), verified_receiver, None);
+
+        // Process a batch that includes a transaction that receives two tokens.
+        let alice = Keypair::new();
+        let tx = Transaction::new(&mint.keypair(), alice.pubkey(), 2, mint.last_id());
+
+        let packets = to_packets(&recycler, &[tx]);
+        verified_sender
+            .send(vec![(packets[0].clone(), vec![1u8])])
+            .unwrap();
+
+        // Process a second batch that spends one of those tokens.
+        let tx = Transaction::new(&alice, mint.pubkey(), 1, mint.last_id());
+        let packets = to_packets(&recycler, &[tx]);
+        verified_sender
+            .send(vec![(packets[0].clone(), vec![1u8])])
+            .unwrap();
+        drop(verified_sender);
+        assert_eq!(banking_stage.join().unwrap(), ());
+
+        // Collect the ledger and feed it to a new bank.
+        let ventries: Vec<_> = entry_receiver.iter().collect();
+
+        // same assertion as below, really...
+        assert_eq!(ventries.len(), 2);
+
+        // Assert the user holds one token, not two. If the stage only outputs one
+        // entry, then the second transaction will be rejected, because it drives
+        // the account balance below zero before the credit is added.
+        let bank = Bank::new(&mint);
+        for entries in ventries {
+            for entry in entries {
+                assert!(
+                    bank.process_transactions(&entry.transactions)
+                        .into_iter()
+                        .all(|x| x.is_ok())
```

Comment:
I see a few things in this test that I'd guess clippy would complain about. Have you looked at it? Or perhaps it doesn't reach our tests?

Reply:
haven't looked at clippy, will do

Reply:
clippy is happy with all these